server/app/services/web-scraper/wss.js

```diff
@@ -0,0 +1,176 @@
+const Queue = require('bull');
+const currency = require('currency.js');
+
+const amazon = require('./amazon-web-scraper');
+const Product = require('../../models/Product');
+
+/* **************************************************************************************** */
+// For testing purposes: REMOVE LATER when putting this file inside the products controller
+// eslint-disable-next-line import/order
+const dotenv = require('dotenv');
+
+const connectDB = require('../../utils/database');
+
+// Load env vars
+dotenv.config({
+  path: './.env',
+});
+
+connectDB();
+/* **************************************************************************************** */
+
+const webScrapeQueue = new Queue('web-scrape-service');
+
+/*  FOR JOBS RECEIVED ON THE WEB SCRAPE SERVICE QUEUE
+    Data for new products:
+    data: {
+      link: LINK_TO_PRODUCT_PAGE,
+      newProduct: true
+    }
+
+    Data for products already in database
+    data: {
+      id: MONGODB_OBJECT_ID (String),
+      link: LINK_TO_PRODUCT_PAGE,
+      currentPrice: CURRENT_PRICE_OF_PRODUCT_IN_DB
+    }
+*/
+
+// Used for when users add new products to database
+const newProductJobOptions = {
+  priority: 1,
+};
+
+const repeat = { every: 20000 }; // Change to a longer duration later
+
+// Used to webscrape existing products in database at a later time
+const repeatJobOptions = {
+  priority: 10,
+  repeat,
+};
+
+webScrapeQueue.process(async job => {
+  // web-scrape link for product details
+  const productDetails = await amazon.getProductDetails(job.data.link);
+
+  const result = {
+    id: job.data.id,
+    productDetails,
+    newProduct: job.data.newProduct,
+  };
+
+  return result;
+});
+
+webScrapeQueue.on('completed', async (job, result) => {
+  const { productDetails } = result;
+
+  // convert currency from string to number
+  productDetails.currentPrice = currency(productDetails.currentPrice).value;
+  productDetails.previousPrice = currency(productDetails.previousPrice).value;
+
+  // if the product is new, save to database and add to queue
+  if (job.data.newProduct) {
+    console.log(
+      `Creating new product in database for ${productDetails.name.substring(
+        0,
+        51,
+      )}\n`,
+    );
+    // TEST - increased currentPrice by $10 so a later fetch will update the price in DB
+    productDetails.currentPrice += 10.0;
+    let product = await Product.create(productDetails);
+    product = product.toJSON();
+
+    await webScrapeQueue.add(
```

Comment:
Does this promise resolve after the job is added or after the job is complete?

Reply:
I tested with the following code
```
const Queue = require('bull');

const myQueue = new Queue('my-queue');

myQueue.process(async job => {
  console.log(`RUNNING JOB with message: ${job.data.msg}`);
});

myQueue.on('completed', (job, result) => {
  console.log('Job Completed');
});

(async () => {
  console.log('Adding job to queue');

  await myQueue.add({ msg: 'TASK1' });

  console.log('Job added to queue');
})();
```
The promise resolves after the job is added. As it currently is written, the ``await`` could be left out. However, the service  would crash if the server attempted to add a message to the queue and Redis is not running. Perhaps I should wrap the 
``await myQueue.add({ msg: 'TASK1' });``
inside a try/catch block. Would that be needed for this project?

Reply:
I think that's a good idea